doc/design/diag_attribute_error.md

```diff
@@ -0,0 +1,96 @@
+# Diagnostic Attribute Error
+
+## Motivation
+
+SQLFlow extended SQL syntax to describe an end-to-end machine learning job, for a typical SQL program:
+
+``` python
+SELECT * FROM iris.train
+TO TRAIN DNNClassifier
+WITH
+  model.n_classes = 2,
+  model.hidden_units = [128, 64]
+LABEL class
+INTO my_model;
+
+SELECT * FROM iris.test
+TO PREDICT iris.pred.class
+USING my_model;
+```
+
+SQLFlow compiles the above SQL program into an execution plan and runs it.  As the TRAIN statement above, SQLFlow uses `TO TRAIN CLAUSE`  to train a specific model called  `DNNClassifier`, using `WITH CLAUSE` to configure the training arguments. 
+
+Sometimes users may make some configuration mistake on `WITH CLAUSE`, then the job would fault during execution and return some uncertain error message.
+
+This documentation issued a way that adding extra contract in [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) to contract the arguments, and this can achieve three advantage at least:
+
+1. Early testing, we can do early testing before running the job; users can wait less time and cluster save resources.
+2. More accurate diagnostic message.
+3. Using docstring is Keras or Tensorflow native; users don't need to modify the model code.
+
+## Design
+
+The docstring of a function or a class can include argument documentation and contracts.
+
+The argument documentation starts with argument name and description followed by a colon `:` and the contrast on this line starts with `#`. The contract should be Python code and return `True` or `False`.
+
+An example:
+
+```python 
+class MyDNNClassifier(keras.Model)
+    def __init__(self, n_classes=32, hidden_units=[32, 64]):
+    """
+    Args:
+
+    # isintance(n_classes, int) && n_classes > 1
```

Comment:
This is a line of code, not some hints. I would recommend we follow the format of: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html, e.g.

```python
class ExampleError(Exception):
    """Exceptions are documented in the same way as classes.

    The __init__ method may be documented in either the class level
    docstring, or as a docstring on the __init__ method itself.

    Either form is acceptable, but the two should not be mixed. Choose one
    convention to document the __init__ method and be consistent with it.

    Note:
        Do not include the `self` parameter in the ``Args`` section.

    Args:
        msg (str, required): Human readable string describing the exception.
        code (:obj:`int`, optional): Error code.
        init_val (int, optional, (0, 1000]): initial value.

    Attributes:
        msg (str): Human readable string describing the exception.
        code (int): Exception error code.

    """
```

Reply:
> This is a line of code, not some hints. I would recommend we follow the format of: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html, e.g.
> 
> ```python
> class ExampleError(Exception):
>     """Exceptions are documented in the same way as classes.
> 
>     The __init__ method may be documented in either the class level
>     docstring, or as a docstring on the __init__ method itself.
> 
>     Either form is acceptable, but the two should not be mixed. Choose one
>     convention to document the __init__ method and be consistent with it.
> 
>     Note:
>         Do not include the `self` parameter in the ``Args`` section.
> 
>     Args:
>         msg (str, required): Human readable string describing the exception.
>   ```
A critical flaw of this method is that we never know what the message is, hence we'll have no control of the user experience. For example:
```python
Args: 
    n_classes (int, required, [2, )): should be >=20
```
In contrast, #2245 has a solid mathematical theory basis. If we really have to parse requirements in the docstrings. A contributor can write:
##### Snippet 1
```python
"""
    :param: n_classes: (Describe the purpose of `n_classes`)
    :require: Int & Greater(1)   
"""
```
This would guarantee the consistency of what the author wants and what the users think what the author wants.
The snippet `Int & Greater(1)` will be substituted by `contracts` to the human-readable form. Therefore, the end-users will never see `Int & Greater(1)`, but `int that's > 1`.

Moreover, the messages generated by `contracts.py` are actually DNFs in plain english, we can easily parse them back to DNFs. With the help of `contracts.py` (and a simple parser implemented in `split`), a contributor can even write:
##### Snippet 2
```python
"""
    :param: n_classes: (Describe the purpose of `n_classes`)
    :require: int that's >= 1   
"""
```
Frankly speaking, if you believe writing requirements in docstring is a good idea, I think **Snippet 1** is enough.

Reply:
> ```python
> """
>     :param: n_classes: (Describe the purpose of `n_classes`)
>     :require: int that's >= 1   
> """
> ```

Do you @shendiaomo  propose some special format for the "contract" description? I think the format like `:param:` is delicate and fragile.  How if the users forgot to type a colon `:`? Do we need another layer of contract to describe this contract?

I understand @typhoonzero 's point is that we need compiler readable checking rules and human-readable explanation.  For example, `isintance(n_classes, int) && n_classes > 1` is not enough -- we also need `n_class must be an integer and larger than 1`.

It looks to me that @Yancey1989 already put this under consideration -- each parameter has a documentation of two lines:

- A line starting with `#` is the compiler-readable checking rule

- The next line (optional) is the human-readable rule.

I think @Yancey1989 's design is very complete.  Did I miss something? @typhoonzero @shendiaomo

Reply:
From @typhoonzero 's comment
>         init_val (int, optional, (0, 1000]): initial value.
This case is very simple if we want to express more complex rules such as `x should be an int list and each element greater then 1`, we should extend the syntax.

As the discussion on the daily meeting, to design a Python library instead of a native Python code to check the parameters can make the code shorter and simpler, but we should collective many user cases and then writing a new design of this Python library in the future.